### DISCUSSION TITLE
Mbed CE Roadmap

### DISCUSSION THREAD

**Discussion by multiplemonomials** [mbed-ce/mbed-os#298] [General]
Hi all!  With the official EoL announcement from Mbed OS, I expect that more people will be attempting to try out and (hopefully) contribute to Mbed CE soon!  So, I'd like to share some details about where I think that Mbed CE can & will go over the next 6-12 months.  For the warm fuzzies, I'll also include some highlights of the progress we've already made in the first two years (wow! two years already! how the time does fly).

# Major Efforts
## CI System
This project's goal is to create an equivalent for Mbed CE to ARM's Mbed device farm and continuous integration system.  This will let us test incoming PRs automatically, and hopefully with more rigor than even ARM had.

✅ [CI board](https://github.com/mbed-ce/mbed-ce-ci-shield-v2) designed, enabling automatic testing for a variety of different interfaces that Mbed supports
🔵 CI board testing is in progress.  V2 R2 board is working pretty well!  Currently I2C master, I2C slave, SPI master, PWM, and ADC tests are nearly good to go but I haven't started on UART or SPI slave tests.
🔵 Setup is also in progress (by @VictorWTang) for a Github Actions-based runner system that can test on real hardware via a device farm
⚪ Once we have a verified working design for the boards, we can build a larger number, potentially via a PCB assembly service!
⚪ After that, the main job will be building out the device farm to include representative devices from all sorts of target families.

## Documentation
Efforts to produce better documentation of the features and APIs available in Mbed.  Official documentation had always been a mixed bag, and I'd really like to improve that state of affairs.

✅ Set up CI job to publish Doxygen documentation to a [public webpage](https://mbed-ce.github.io/mbed-os/group__mbed-os-public.html)
✅ Write basic webpage generator to publish a list of [supported targets](https://mbed-ce.github.io/mbed-ce-test-tools/targets/), [features/drivers](https://mbed-ce.github.io/mbed-ce-test-tools/drivers/), and [test runs](https://mbed-ce.github.io/mbed-ce-test-tools/tests/)
⚪ Webpage generator could still use lots more polish, such as importing content written for the wiki into the generated webpages.
⚪ Still need to set up some kind of pipeline to publish the handwritten, non-Doxygen docs (https://github.com/ARMmbed/mbed-os-5-docs).  

## Arduino IDE Support

I stand by my decisions regarding Mbed CE's build system changes.  Embracing CMake fully, without any Python wrappers, makes programming with Mbed much more similar to standard C++ projects.  Additionally, making Mbed CE primarily target CLion and VS Code as IDEs for development (plus a command line option as well) introduces a lot more developer freedom and lets folks take advantage of battle-tested dev tools.

However, with the impending shutdown of Mbed Studio and Keil Studio Cloud (I think?), we also need an IDE option that requires little to no up-front setup, even if it ends up being a bit more limited in terms of the use cases it supports.  This is especially important for teachers and students who use Mbed in class, as it's not desirable to spend lots of time and effort setting up the toolchain on each student's computer.

To this end, Mbed CE is working towards becoming compatible with the Arduino IDE.  This compatibility would mean that you could install support for Mbed board through the Arduino library manager, and ideally they would work just like any other Arduino board.  Since Mbed OS is already used as part of a number of arduino boards, this is well on its way to being a supported option.

✅ Contact Arduino and understand process for becoming listed in the Arduino package management system (we have to host a JSON index file and send them the URL)
✅ Create [system](https://github.com/mbed-ce/ArduinoCore-mbed-ce) to export Mbed compiler flags & compiled libraries in a format that the Arduino IDE can understand
🔵 Create CI job to automatically build Arduino cores for each board and publish an index of the compiled cores
⚪ Create a way to add even Mbed board that aren't arduinos into this system (as in, some way to auto generate or stub out the needed header files without manually writing them for each target)
⚪ Publish libraries in the Arduino IDE index

## New Targets
Target devices that Mbed CE has added support for

✅ Teensy 4.0 and 4.1 supported, and MIMXRT1060 EVK as well (via extending the preexisting MIMXRT1050 MCU support)
✅ Arduino Nano 33 BLE supported (via existing nRF52840 support)
✅ Arduino Nicla Sense ME supported (via existing nRF52833 support)
✅ Raspberry Pi Pico supported (completely new vendor & device to Mbed, adapted & completed the in-progress PR from Arduino)
✅ STM32H5 family port (basic features working, still missing USB and some other drivers though) (by @JohnK1987 and @multiplemonomials)
🔵 Arduino Giga & Portenta in the process of being added (by @multiplemonomials)
⚪ Arduino Nicla Vision board was donated by Arduino and will be ported at some point

## Improvements and Bugfixes
Grab bag of sorely needed bugfixes and improvements that Mbed CE made and is making to Mbed OS.

✅ Removed tens of obsolete & unbuyable target boards from the Mbed codebase
✅ Significantly extended min & max PWM frequency limits on RPi Pico and all STM32 targets -- you can now have frequencies of up to 1MHz and as as low as 10Hz.
✅ Enabled floating point printf by default and increased default serial baudrate to 115200
✅ Fixed lots of non-API-compliant SPI & I2C behavior on LPC1768 and STM32 targets
✅ Added [DMA SPI support](https://github.com/mbed-ce/mbed-os/wiki/STM32-DMA-SPI-Support) for STM32 targets
✅ Wrote pages of new documentation for the [SPI](https://mbed-ce.github.io/mbed-os/classmbed_1_1_s_p_i.html) and [I2C](https://mbed-ce.github.io/mbed-os/classmbed_1_1_i2_c.html) classes
✅ Updated the JSON parsing code to allow writing config files in JSON5, with comment support!
✅ Working on restoring the [memory bank information](https://github.com/mbed-ce/mbed-os/wiki/Mbed-Memory-Bank-Information) features from Mbed CLI 1
✅ New [CompositeEMAC Ethernet driver API](https://github.com/mbed-ce/mbed-os/blob/master/connectivity/drivers/emac/CompositeEMAC.md) added
✅ STM32H7 Ethernet driver rewritten from the ground up using CompositeEMAC
✅ Update the build system to build for either mbed-baremetal or mbed-os, not both.  The current build system compiles most .a libraries against baremetal but then can link them with baremetal or mbed-os, which causes some subtle issues (e.g. linker errors with cellular).  This should be fixed by adding a new CMake or JSON option (or both?) to select whether baremetal or Mbed OS is wanted for a project.
⚪ Add a system for defining "capabilities" of different MCUs and boards, e.g. "what are the max and min PWM frequencies this MCU can support" or "are zero-length I2C transactions supported" or "does async SPI support DMA", etc.

Anyway, I hope that this provides a bit more of a look into what we are working on at Mbed CE, and what this project has already managed to do so far!

And remember, Mbed CE is a _community_ edition for a reason!  We are committed to accepting PRs from the community, and the more help we have, the more we can make Mbed OS into what it always should have been.  If you have any suggestions, please leave a comment here or create an issue or a PR!  Or, if you are interested in maintaining some part of Mbed, *please* let us know!  We badly need maintainers for lots of the device families, and we really need people familiar with Bluetooth, USB, and other protocols used by Mbed!

**Comment by noutram-old-uopaccount** [comment]
This is great to read. I hope to contribute. Let's hope the EOL announcement is a new beginning.

**Reply by multiplemonomials** [reply]
Awesome, thank you for getting in touch!  Anything specific that you are interested in contributing to?  Believe me, we need work in lots of different areas!

**Comment by avdept** [comment]
Just read email this morning and although I wasnt active user of Mbed for few years it was sad to see its EoL. But when I got to the end of email with mention on Mbed CE - that was such a relief! Glad this project wasnt fully abandoned, so keep it up 👍🏻

**Comment by rlanders** [comment]
This is great stuff! I actually have been in contact with Reinhard Keil regarding MbedOS over the past years, and I was imploring him to turn Mbed in to a community supported thing like Zephyr, since it was clear that ARM wasn't going to spend another penny on it. I doubt he really listened to me, but I think I got what I wanted here. I have been in the bowls of this OS, and have made customizations for my needs in the past and am more than willing to do that here. My immediate thought was to streamline the OS to be more like the "bare metal" flavor by default, and then allow the user to include the features as they are needed. This is more in line with how Arduino works, and would speed up first builds significantly (which happens every time you change mbed_app.json)

**Reply by rlanders** [reply]
OK, I just did an initial compile of the hello world project, and WOW. You definitely delivered on a much better first-time compile time, and I also see that the resulting binary is about 15k smaller than the mbedOS v5.17 version of that project. I never update to mbed v6 because it broke enough things in my code that it just wasn't worth it.

Does there happen to be a porting guide? If not, can there be?

**Reply by multiplemonomials** [reply]
That's awesome!  Thank you!  Glad to hear it's working out!  We would be happy to accept contributions for making the bare metal version of Mbed even more bare metal!

Actually, the effort to make Mbed smaller was started by ARM themselves!  With Mbed CLI 2, they started the process of breaking the monolithic library down into a smaller set of CMake targets -- the main `mbed-os` and `mbed-baremetal` targets now just build the core HAL and drivers, and all of the networking, storage, bluetooth, etc functionality is split into its own linkable targets, such as `mbed-netsocket` and `mbed-ble`.  This really, really helped with the huge compile time issues that affected Mbed 6.

I would be happy to help you port stuff over!  Are you using Mbed on a custom target, or on a target that was supported but was removed?  Or do you have other extensions to Mbed that you'd need to work in?  We have a documented example for custom targets available here: https://github.com/mbed-ce/mbed-ce-custom-targets

**Reply by rlanders** [reply]
I actually have a few custom targets, which happen to be modules based on the nRF52840. Basically, due to lack of parts, I was forced to redesign the board a couple times. I will take a look at the custom targets example for sure. With respect to porting assistance, I would have to get approval from the investors, but I wouldn't mind talking further.

**Comment by Kalinzon** [comment]
Hello!
I think you know what you're doing guys. But maybe you'll try to integrate Mbed CE into the Platformio plugin/framework instead of Arduino iDE? It is really more powerful, and inegrates with both Clion and VSCode.

**Reply by multiplemonomials** [reply]
That's a really good point.  I myself have not used platformio and I don't know what it would take to get Mbed integrated with it, but I've heard good things about it.  It would be really awesome if someone could take on the task of integrating it!

**Reply by horeich** [reply]
We've been using PlatformIO for a while together with mbedOS (VS Code extension). It's really handy and easy to use and I think we should push that it stays supported by PIO.

**Comment by ndrs-pst** [comment]
@multiplemonomials Just an Idea, if `Mbed CE` is repositioned as an intermediate layer on top of `Zephyr RTOS`, it might be more appropriate to rebrand it with a name that reflects its new role. How about `Mbed RTE` which stand for `Runtime Environment` ?

**Reply by JohnK1987** [reply]
MbedOS is based on ARM/Keil RTX RTOS and has nothing together with Zephyr. Mbed CE (Community Edition) is github community organization as substitute for ARM Mbed github organization because ARM stoped development and support.

**Reply by ndrs-pst** [reply]
How about this direction (https://blog.arduino.cc/2024/07/24/the-end-of-mbed-marks-a-new-beginning-for-arduino/), 
but with `Mbed` itself on top of Zephyr?
I highly value `Mbed OS`'s APIs, like drivers, RTOS, storage, and connectivity 😃.

**Reply by Kalinzon** [reply]
Man, Mbed itself is an RTOS. Why mount an RTOS over an RTOS?

**Reply by ndrs-pst** [reply]
My idea is about transforming Mbed from a standalone RTOS into Run Time Environment (RTE) that works in conjunction with Zephyr. This way, the task of supporting new targets can be handed over to Zephyr instead.

**Reply by JohnK1987** [reply]
Yeah, good idea and when you start? :)

**Comment by mcuxmx** [comment]
Does mbed-ce have plans to support more processors, such as risc-v?

**Reply by JohnK1987** [reply]
Hello,

Be so kind read https://github.com/mbed-ce/mbed-os/discussions/303

**Comment by horeich** [comment]
I wonder what's the best way to contribute drivers for external ICs connected by I2C, SPI, UART etc. We have many of those and like to contribute to keep the project going. Does it make sense to add those inside this repo and adding them using e.g. COMPONENT_ADD or should these provided as external repos?

**Reply by JohnK1987** [reply]
Hello, for libraries we have another organization - https://github.com/mbed-ce-libraries-examples

**Reply by horeich** [reply]
Ok, thanks! And how would I add a library there. Should I start a discussion?

**Reply by JohnK1987** [reply]
Yes, would be nice.
https://github.com/mbed-ce-libraries-examples/.github/blob/main/profile/README.md#how-to-contribute
Then we will discuss how to proceed this.

If you will face some issue with access let know.
